### PR TITLE
gossip: retry bootstrapping if we can't start a client

### LINKED ISSUE #4714: circleci: failed tests (13504): TestPut
The following test appears to have failed:

[#13504](https://circleci.com/gh/cockroachdb/cockroach/13504):

```
I160229 16:00:28.841635 acceptance/replication_test.go:66  found 2 replicas
I160229 16:00:29.843953 acceptance/replication_test.go:66  found 2 replicas
I160229 16:00:30.846499 acceptance/replication_test.go:66  found 2 replicas
I160229 16:00:31.849286 acceptance/replication_test.go:66  found 2 replicas
I160229 16:00:31.950534 acceptance/cluster/localcluster.go:606  stopping
--- FAIL: TestPut (30.76s)
	testing.go:188: acceptance/replication_test.go:72: condition failed to evaluate within 20s: expected 3 replicas, only found 2
=== RUN   TestDockerPython
I160229 16:00:33.043743 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
I160229 16:00:33.790377 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
I160229 16:00:35.362635 acceptance/cluster/localcluster.go:262  started docker-spy: 172.17.0.50
I160229 16:00:35.362680 acceptance/cluster/localcluster.go:269  Initializing Cluster:
Cluster Setup: AdHoc 3x1
Duration: 5s, Stall: 2m0s
Node 0 - Store 0
Node 1 - Store 0
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:30: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	util_test.go:64: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

### LINKED ISSUE #4822: circleci: failed tests (13717): TestDockerC
The following test appears to have failed:

[#13717](https://circleci.com/gh/cockroachdb/cockroach/13717):

```
  cockroach: [start --certs=/certs --host=roach2.local --port=26257 --log-threshold=INFO --scan-max-idle-time=200ms --store=/data3.0 --join=roach0.local:26257 --log-dir=/logs/roach2.local --logtostderr=false --alsologtostderr=true]
I160303 17:12:52.820100 acceptance/replication_test.go:50  waiting for first range to have 3 replicas
I160303 17:12:53.821607 acceptance/replication_test.go:66  found 2 replicas
I160303 17:13:10.573656 acceptance/replication_test.go:66  found 2 replicas
I160303 17:13:10.674187 acceptance/cluster/localcluster.go:596  stopping
--- FAIL: TestDockerC (37.54s)
	testing.go:170: acceptance/replication_test.go:72: condition failed to evaluate within 15s: expected 3 replicas, only found 2
=== RUN   TestClusterRecovery
--- SKIP: TestClusterRecovery (0.00s)
	chaos_test.go:300: TODO(pmattis): #4517
=== RUN   TestNodeRestart
--- SKIP: TestNodeRestart (0.00s)
	chaos_test.go:360: TODO(pmattis): #4517
=== RUN   TestGossipPeerings
I160303 17:13:11.494625 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
I160303 17:13:11.690557 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:30: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	util_test.go:63: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

### LINKED ISSUE #4835: circleci: failed tests (13747): TestDockerPython
The following test appears to have failed:

[#13747](https://circleci.com/gh/cockroachdb/cockroach/13747):

```
  cockroach: [start --certs=/certs --host=roach2.local --port=26257 --log-threshold=INFO --scan-max-idle-time=200ms --store=/data3.0 --join=roach0.local:26257 --log-dir=/logs/roach2.local --logtostderr=false --alsologtostderr=true]
I160303 20:57:16.188578 acceptance/replication_test.go:50  waiting for first range to have 3 replicas
I160303 20:57:17.192149 acceptance/replication_test.go:66  found 2 replicas
I160303 20:57:32.520595 acceptance/replication_test.go:66  found 2 replicas
I160303 20:57:32.621118 acceptance/cluster/localcluster.go:596  stopping
--- FAIL: TestDockerPython (28.10s)
	testing.go:170: acceptance/replication_test.go:72: condition failed to evaluate within 15s: expected 3 replicas, only found 2
=== RUN   TestDockerRuby
I160303 20:57:33.514135 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
I160303 20:57:34.999317 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
I160303 20:57:36.230809 acceptance/cluster/localcluster.go:261  started docker-spy: 172.17.0.55
I160303 20:57:36.230847 acceptance/cluster/localcluster.go:268  Initializing Cluster:
Cluster Setup: AdHoc 3x1
Duration: 5s, Stall: 2m0s
Node 0 - Store 0
Node 1 - Store 0
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:30: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	util_test.go:63: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

### LINKED ISSUE #4892: circleci: failed tests (13896): TestGossipPeerings TestDockerRuby
The following test appears to have failed:

[#13896](https://circleci.com/gh/cockroachdb/cockroach/13896):

```
I160306 00:42:23.518269 acceptance/replication_test.go:66  found 1 replicas
I160306 00:42:24.520021 acceptance/replication_test.go:66  found 1 replicas
I160306 00:42:25.521786 acceptance/replication_test.go:66  found 1 replicas
I160306 00:42:26.523521 acceptance/replication_test.go:66  found 1 replicas
I160306 00:42:26.624140 acceptance/cluster/localcluster.go:596  stopping
--- FAIL: TestGossipPeerings (19.68s)
	testing.go:170: acceptance/replication_test.go:72: condition failed to evaluate within 15s: expected 3 replicas, only found 1
=== RUN   TestGossipRestart
--- SKIP: TestGossipRestart (0.00s)
	gossip_peerings_test.go:147: TODO(pmattis): #4445
=== RUN   TestDockerJava
I160306 00:42:28.879205 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
....I160306 00:42:29.662730 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
I160306 00:42:30.067953 acceptance/cluster/localcluster.go:261  started docker-spy: 172.17.0.3
I160306 00:42:30.068002 acceptance/cluster/localcluster.go:268  Initializing Cluster:
Cluster Setup: AdHoc 3x1
--
I160306 00:43:42.454459 acceptance/replication_test.go:66  found 1 replicas
I160306 00:43:43.456208 acceptance/replication_test.go:66  found 1 replicas
I160306 00:43:44.457907 acceptance/replication_test.go:66  found 1 replicas
I160306 00:43:45.459766 acceptance/replication_test.go:66  found 1 replicas
I160306 00:43:45.560352 acceptance/cluster/localcluster.go:596  stopping
--- FAIL: TestDockerRuby (20.12s)
	testing.go:170: acceptance/replication_test.go:72: condition failed to evaluate within 15s: expected 3 replicas, only found 1
=== RUN   TestSingleKey
I160306 00:43:48.021727 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
....I160306 00:43:48.414400 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
I160306 00:43:48.943010 acceptance/cluster/localcluster.go:261  started docker-spy: 172.17.0.3
I160306 00:43:48.943055 acceptance/cluster/localcluster.go:268  Initializing Cluster:
Cluster Setup: AdHoc 3x1
Duration: 5s, Stall: 2m0s
Node 0 - Store 0
Node 1 - Store 0
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:30: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	util_test.go:66: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

### LINKED ISSUE #4893: circleci: failed tests (13898): TestBuildInfo TestDockerPython
The following test appears to have failed:

[#13898](https://circleci.com/gh/cockroachdb/cockroach/13898):

```
I160306 00:59:26.896842 acceptance/replication_test.go:66  found 1 replicas
I160306 00:59:27.898662 acceptance/replication_test.go:66  found 1 replicas
I160306 00:59:28.900287 acceptance/replication_test.go:66  found 1 replicas
I160306 00:59:29.902096 acceptance/replication_test.go:66  found 1 replicas
I160306 00:59:30.002882 acceptance/cluster/localcluster.go:596  stopping
--- FAIL: TestBuildInfo (61.15s)
	testing.go:170: acceptance/replication_test.go:72: condition failed to evaluate within 15s: expected 3 replicas, only found 1
=== RUN   TestDockerC
I160306 00:59:32.264715 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
....I160306 00:59:32.525254 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
I160306 00:59:32.825828 acceptance/cluster/localcluster.go:261  started docker-spy: 172.17.0.3
I160306 00:59:32.825877 acceptance/cluster/localcluster.go:268  Initializing Cluster:
Cluster Setup: AdHoc 3x1
Duration: 5s, Stall: 2m0s
Node 0 - Store 0
Node 1 - Store 0
--
I160306 01:01:21.302831 acceptance/replication_test.go:66  found 1 replicas
I160306 01:01:22.304642 acceptance/replication_test.go:66  found 1 replicas
I160306 01:01:23.306624 acceptance/replication_test.go:66  found 1 replicas
I160306 01:01:24.308258 acceptance/replication_test.go:66  found 1 replicas
I160306 01:01:24.408832 acceptance/cluster/localcluster.go:596  stopping
--- FAIL: TestDockerPython (28.51s)
	testing.go:170: acceptance/replication_test.go:72: condition failed to evaluate within 15s: expected 3 replicas, only found 1
=== RUN   TestDockerRuby
I160306 01:01:26.969492 acceptance/cluster/docker.go:72  ImagePull cockroachdb/docker-spy:20160209-143235 starting
....I160306 01:01:27.567760 acceptance/cluster/docker.go:87  ImagePull cockroachdb/docker-spy:20160209-143235 complete
I160306 01:01:28.222274 acceptance/cluster/localcluster.go:261  started docker-spy: 172.17.0.3
I160306 01:01:28.222321 acceptance/cluster/localcluster.go:268  Initializing Cluster:
Cluster Setup: AdHoc 3x1
Duration: 5s, Stall: 2m0s
Node 0 - Store 0
Node 1 - Store 0
--
--- SKIP: TestBuildBabyCluster (0.00s)
	terraform_test.go:30: only enabled during testing
=== RUN   TestFiveNodesAndWriters
--- SKIP: TestFiveNodesAndWriters (0.00s)
	util_test.go:66: running in docker mode
FAIL
ok github.com/cockroachdb/cockroach/acceptance 1337s

```
Please assign, take a look and update the issue accordingly.

### LINKED ISSUE #4894: circleci: docker flakiness
Every now and then (but often enough to fail a significant amount of builds due to the sheer volume of iterations this bug gets), Gossip doesn't join up:
```
I160306 00:51:43.699152 cli/start.go:129  [build] alpha.v1-676-g3325b63 @ 2016/03/06 00:37:29 (go1.6)
I160306 00:51:43.701589 server/context.go:223  1 storage engine initialized
I160306 00:51:43.701841 cli/start.go:143  starting cockroach node
E160306 00:51:43.896722 gossip/gossip.go:733  invalid bootstrap address: &{typ:tcp addr:roach0.local:26257 exhausted:false}, lookup roach0.local: no such host
I160306 00:51:43.900384 storage/engine/rocksdb.go:106  opening rocksdb instance at "/data2.0"
I160306 00:51:43.909866 server/node.go:353  store store=0:0 ([]=/data2.0) not bootstrapped
I160306 00:51:43.910160 storage/stores.go:289  read 0 node addresses from persistent storage
I160306 00:51:43.911156 server/node.go:486  connecting to gossip network to verify cluster ID...
```

Note `E160306 00:51:43.896722 gossip/gossip.go:733  invalid bootstrap address: &{typ:tcp addr:roach0.local:26257 exhausted:false}, lookup roach0.local: no such host` which suggests that there's maybe a DNS issue with `docker-spy` and a subsequent failure in Gossip to retry in a speedy manner.

Repros locally, for example

```
while [ $? -eq 0 ]; do rm -rf ./acceptance/Test*; make acceptance TESTS=GossipPeerings TESTFLAGS='-v -l .'; done
```

### PR DESCRIPTION
**Pull request by petermattis**
Fixes #4714.
Fixes #4822.
Fixes #4835.
Fixes #4892.
Fixes #4893.
Fixes #4894.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4895)

### CODE REVIEW

**Comment by tamird** [comment]
Is it possible to add a unit test for this?

**Comment by tschottdorf** [comment]
LGTM. What a combo. Prefer not to block this on a test unless it's trivial.

**Comment by petermattis** [comment]
@tamird How about I file a bug to add a unit test?

**Comment by tamird** [comment]
Works for me. LGTM.

File: gossip/gossip.go [hunk 1]
```diff
@@ -777,14 +781,14 @@ func (g *Gossip) bootstrap() {
 			select {
 			case <-bootstrapTimer.C:
 				bootstrapTimer.Read = true
-				// continue
+				// break
```
Comment (bdarnell) [inline gossip/gossip.go]: What's going on with these comments? They should probably be removed, but if they were ever uncommented then it looks like the intention is to continue, not break.

Comment (petermattis) [inline gossip/gossip.go]: I don't think the intention was to continue in this case as then we would never reach the second select statement. Perhaps the original author was trying to document the control flow and was indicating "continue with the rest of the loop".

Comment (spencerkimball) [inline gossip/gossip.go]: The intention wasn't to continue. Bad choice of words. "continue" just meant proceed to next select statement.

**Pull request merged**